### PR TITLE
deal.II: Update to version 9.4.0

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

**Review by luca-heltai — CHANGES_REQUESTED** [review]
We also need 

```
    conflicts('cxxstd=default', when='@9.4:+cgal')
    conflicts('cxxstd=14', when='@9.4:+cgal')
```
since CGAL supports requires cxx17 support.

**Comment by jppelteret** [comment]
I think that conflicting with `cxxstd=default` might be a bit too strict, because with a sufficiently modern compiler the default might be to use C++17:
```bash
$ g++ -dM -E -x c++  /dev/null | grep -F __cplusplus                                                                                                                                                  
#define __cplusplus 201703L

$ g++ -v                                                                                                                                                                                              
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/12.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror --with-build-config=bootstrap-lto --enable-link-serialization=1
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (GCC) 
```
I can confirm that I managed to get past the configuration phase without considering the `cxxstd` that has been set (and CGAL itself was successfully built with this compiler).

I'll mark the incompatibility with `C++14` though. Thanks very much for the recommendation!

**Comment by luca-heltai** [comment]
I've tried this patch on an intel mac, and cxx17 is not picked up, even forcing cxxstd=17. I'll experiment again tomorrow, but we might have an issue (not related to this patch though).

**Comment by luca-heltai** [comment]
Did deal.II pass the configuration phase with this patch for you, with cgal enabled? It did not on my mac. I had to manually add `CMAKE_CXX_FLAGS="-std=c++17"` for the configure to pick up cxx17.

**Review by luca-heltai — DISMISSED** [review]

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by jppelteret** [comment]
Hi @luca-heltai ,

> Did deal.II pass the configuration phase with this patch for you, with cgal enabled? It did not on my mac. I had to manually add CMAKE_CXX_FLAGS="-std=c++17" for the configure to pick up cxx17.

Yeah, configuration worked for me (I have problems building the library, but that seems to be a separate issue). Details below. I had previously had success with limiting the C++ standard by setting these flags. Looking again at what I [previously implemented](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/dealii/package.py#L382-L386) and comparing it to [our CMake docs](https://dealii.org/developer/users/cmake_dealii.html#configure), it looks like I got it a bit wrong. I'll add a commit to try to adjust this.

```
$ ./spack spec -Il dealii%gcc@12.1.0~hdf5
Input spec
--------------------------------
 -   dealii%gcc@12.1.0~hdf5

Concretized
--------------------------------
 -   ofcel4t  dealii@9.4.0%gcc@12.1.0+adol-c+arborx+arpack+assimp+cgal~cuda~doc+examples+ginkgo+gmsh+gsl~hdf5~int64~ipo+metis+mpi+muparser~nanoflann~netcdf+oce~optflags+p4est+petsc~python+scalapack+simplex+slepc+sundials+symengine+threads+trilinos build_type=DebugRelease cxxstd=default arch=linux-manjaro21-zen
```
(Note that this sets `cxxstd=default` by default)

```
-- Performing Test DEAL_II_HAVE_CXX17_FEATURES
-- Performing Test DEAL_II_HAVE_CXX17_FEATURES - Success
-- Performing Test DEAL_II_HAVE_CXX17_CONSTEXPR_LAMBDA_BUG_OK
-- Performing Test DEAL_II_HAVE_CXX17_CONSTEXPR_LAMBDA_BUG_OK - Success
-- C++17 support is enabled.
-- Performing Test DEAL_II_HAVE_CXX20_FEATURES
-- Performing Test DEAL_II_HAVE_CXX20_FEATURES - Failed
```

```
-- Include <PATH>/spack-stage/spack-stage-dealii-9.4.0-ofcel4tzda6gistwytybmdz3gufojjpm/spack-src/cmake/configure/configure_50_cgal.cmake
-- Boost include dirs: <PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/include
-- Boost libraries:    <PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_iostreams-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_serialization-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_system-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_thread-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_regex-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_chrono-mt.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/boost-1.79.0-j2z4vtz6amdaeckkcymyxshm4n2qmlf6/lib/libboost_atomic-mt.so
--   CGAL_VERSION: 5.0.3
--   CGAL_INCLUDE_DIRS: <PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/cgal-5.0.3-4k5cziizwccmrhzagkna2epo654mkn76/include
--   CGAL_USER_INCLUDE_DIRS: <PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/cgal-5.0.3-4k5cziizwccmrhzagkna2epo654mkn76/include
--   CGAL_LIBRARIES: <PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/mpfr-4.1.0-e67bnxnfn72sntz22qmv4pnnpwdpqtod/lib/libmpfr.so;<PATH>/opt/spack/linux-manjaro21-zen/gcc-12.1.0/gmp-6.2.1-oznbtkvyzat447b5kcm6grdnf2uulvuv/lib/libgmp.so
-- Found CGAL
-- Performing Test DEAL_II_CGAL_HAS_DEPRECATED_BOOST_INCLUDES_OK
-- Performing Test DEAL_II_CGAL_HAS_DEPRECATED_BOOST_INCLUDES_OK - Failed
-- Test unsuccessful, define DEAL_II_CGAL_HAS_DEPRECATED_BOOST_INCLUDES
-- DEAL_II_WITH_CGAL successfully set up with external dependencies.
```

> I've tried this patch on an intel mac, and cxx17 is not picked up, even forcing cxxstd=17. I'll experiment again tomorrow, but we might have an issue (not related to this patch though).

I must admit that I never got C++17 working on my Mac, even when I tried to build with a modern compiler. But maybe that's also due to the implementation error :-/

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -289,6 +296,13 @@ class Dealii(CMakePackage, CudaPackage):
               msg='The interface to Nanoflann was removed from version 9.3.0. '
                   'Please explicitly disable this variant via ~nanoflann')
 
+    # interfaces added in 9.4.0:
+    for p in ['cgal']:
+        conflicts('+{0}'.format(p), when='@:9.3',
+                  msg='The interface to {0} is supported from version 9.4.0 '
+                      'onwards. Please explicitly disable this variant '
+                      'via ~{0}'.format(p))
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dealii/package.py]: Can you remove this conflict and use a conditional variant?

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -73,6 +74,8 @@ class Dealii(CMakePackage, CudaPackage):
             description='Compile with Arpack and PArpack (only with MPI)')
     variant('adol-c',   default=True,
             description='Compile with ADOL-C')
+    variant('cgal',     default=True,
+            description='Compile with CGAL')
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dealii/package.py]: Variant can also be conditional:
```suggestion
    variant('cgal', default=True, when='@9.4:',
            description='Compile with CGAL')
```

**Review by alalazo — CHANGES_REQUESTED** [review]
One minor request

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Ah, great! Thanks for the suggestion. I don't think that I (we?) knew about this. Maybe I should update the other variants in a follow up PR.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 1 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Done. Thank you!

**Review by alalazo — APPROVED** [review]

**Pull request merged**